src/cluster.c

```diff
@@ -5837,11 +6066,22 @@ clusterNode *getNodeByQuery(client *c, struct redisCommand *cmd, robj **argv, in
                 /* If it is not the first key, make sure it is exactly
                  * the same key as the first we saw. */
                 if (!equalStringObjects(firstkey,thiskey)) {
-                    if (slot != thisslot) {
+                    /* If the channel exists across different slots, it is still
+                     * valid if they are being served from the same node. */
+                    if (is_pubsublocal) {
+                        if (n != thisnode) {
+                            getKeysFreeResult(&result);
+                            if (error_code) {
+                                *error_code = CLUSTER_REDIR_CROSS_NODE_CHANNEL;
+                            }
+                            return NULL;
+                        }
+                    } else if (slot != thisslot) {
                         /* Error: multiple keys from different slots. */
                         getKeysFreeResult(&result);
-                        if (error_code)
+                        if (error_code) {
                             *error_code = CLUSTER_REDIR_CROSS_SLOT;
```

Comment:
General preference is to not touch code you aren't changing to make git history easier to follow.